### DISCUSSION TITLE
Advice on how to handle the "getContractFactory" augmentations

### DISCUSSION THREAD

**Discussion by PaulRBerg** [ethereum-ts/TypeChain#446] [General]
My [hardhat-packager](https://github.com/paulrberg/hardhat-packager) plugin builds upon `@typechain/hardhat` to prepare the `artifacts` and the `typechain` folders for registry deployment. What it does is to delete all files unrelated to an allowlist of contracts, specified by the user.

I want to add support for TypeChain factories, but I noticed that you aggregate the augmentations for the `getContractFactory` function in one file called `hardhat.d.ts`. This is problematic for me because I would have to edit this file and remove all augmentations that don't refer to the allowlisted contracts.

On my end, I can think of two solutions:

1. Do not include the `hardhat.d.ts` file, but include the bespoke factories from the `factories` folder.
2. Include the `hardhat.d.ts` file as it is, with the downside of the user seeing augmentations for contracts not part of the deployed package (e.g. test contracts).

On your end, I can think of two other solutions:

1. Let us select what contracts we'd like to generate types for.
2. Move the augmentation for the `getContractFactory` function in the bespoke contract declaration file.

Re 2nd solution, taking my [solidity-template](https://github.com/paulrberg/solidity-template) as an example, you would add something like this in the `typechain/Greeter.d.ts` file:

```ts
import { ethers } from "ethers";
import {
  HardhatEthersHelpers as HardhatEthersHelpersBase,
} from "@nomiclabs/hardhat-ethers/types";
import { Greeter__factory } from "./factories/Greeter__factory";

declare module "hardhat/types/runtime" {
  interface HardhatEthersHelpers extends HardhatEthersHelpersBase {
    getContractFactory(
      name: "Greeter",
      signerOrOptions?: ethers.Signer | FactoryOptions
    ): Promise<Greeter__factory>;
  }
}
```

**Comment by PaulRBerg** [comment]
**Update**: I ended up implementing my 1st proposed solution in [hardhat-packager@1.1.0](https://github.com/paulrberg/hardhat-packager/releases/tag/v1.1.0).

**Comment by krzkaczor** [comment]
Interesting problem. I wonder what if there would be an optional glob pattern for typechain plugin ie. rather than generate typings for all artifacts just generate it to these files. Would this help in integration?

**Reply by PaulRBerg** [reply]
Oh this would be fantastic!!
